examples/07_advanced/plot_efficient_data_loading.py

```diff
@@ -0,0 +1,167 @@
+    return np.asarray(load(fmri_path).dataobj)[
+        np.asarray(load(mask_path).dataobj).astype(bool)
+    ]
+
+
+def mask_fmri_single_shared(img, mask_path):
+    return img[np.asarray(load(mask_path).dataobj).astype(bool)]
+
+
+def mask_fmri_parallel(fmri_path, mask_paths, n_jobs=6):
+    fmri_ts = Parallel(n_jobs=n_jobs)(
+        delayed(mask_fmri_single)(fmri_path, mask) for mask in mask_paths
+    )
+    return fmri_ts
+
+
+def mask_fmri_efficient_parallel(fmri_path, mask_paths, n_jobs=6):
+    fmri_ts = Parallel(n_jobs=n_jobs)(
+        delayed(mask_fmri_single_efficient)(fmri_path, mask)
+        for mask in mask_paths
+    )
+    return fmri_ts
+
+
+def mask_fmri_shared_parallel(img, mask_paths, n_jobs=6):
+    fmri_ts = Parallel(n_jobs=n_jobs)(
+        delayed(mask_fmri_single_shared)(img, mask) for mask in mask_paths
+    )
+    return fmri_ts
+
+
+def main(n_images=1, n_regions=6):
+    """
+    Compare the performance of NiftiMasker vs. numpy masking by:
+    1. fetching `n_images` subjects from development fMRI dataset and
+    `n_regions` regions from the Difumo atlas.
+    2. then converting these regions to binary masks.
+    3. then using the NiftiMasker to mask the data.
+    4. then also using dataobj to load the data directly from the nifti file
+    and masking it simply using numpy.
+
+    Parameters
+    ----------
+    n_images : int, default=1
+        Number of subjects to fetch from the development fMRI dataset. These
+        subject images would be concatenated to form a single nifti file.
+        Can be increased to simulate larger data.
+
+    n_regions : int, default=6
+        Number of regions to fetch from the Difumo atlas. These regions would
+        be converted to binary masks and used to mask the fMRI data. This is
+        also the number of jobs to run in parallel.
+    """
+    fmri_path = get_fmri_path(n_subjects=n_images)
+    atlas_path = get_atlas_path()
+    mask_paths = atlas_to_masks(atlas_path, fmri_path, n_regions=n_regions)
+
+    print("waiting")
+    time.sleep(30)
+    print("start nifti masker")
+
+    ts_nifti_masker = mask_fmri_parallel(
+        fmri_path, mask_paths, n_jobs=n_regions
+    )
+    del ts_nifti_masker
+
+    print("waiting")
+    time.sleep(30)
+    print("start efficient masker")
+
+    ts_efficient = mask_fmri_efficient_parallel(
+        fmri_path, mask_paths, n_jobs=n_regions
+    )
+    del ts_efficient
+
+    print("waiting")
+    time.sleep(30)
+    print("load image in shared memory")
+
+    fmri_data = np.asarray(load(fmri_path).dataobj)
+    shm = SharedMemory(create=True, size=fmri_data.nbytes)
+    shared_img = np.ndarray(
+        fmri_data.shape, dtype=fmri_data.dtype, buffer=shm.buf
+    )
+    np.copyto(shared_img, fmri_data)
+    del fmri_data
+
+    print("waiting")
+    time.sleep(30)
+    print("start shared masker")
+    ts_shared = mask_fmri_shared_parallel(
+        shared_img, mask_paths, n_jobs=n_regions
+    )
```

Comment:
Ok so @MengxingLiu 's solution works even better. 

In L147-152 we load the image in a shared memory, which is then shared in L158 across 20 parallel jobs (avoiding loading the data 20 times simultaneously).
![memory_profiling_10_images_20_regions_shared](https://github.com/user-attachments/assets/8d5ac893-b2ee-4b78-bad8-7081a1edb68e)